### PR TITLE
Implementing S3 wrangler service allowing to test connection, list bu…

### PR DESCRIPTION
**Pull request by gsps1**
…ckets, explore S3 buckets and load a file identified by bucket-name and key.

### CODE REVIEW

**Comment by gsps1** [comment]
going to close and re-open, need to re-open against release/3.0. develop has other changes, so changing base doesn't help.

**Comment by gsps1** [comment]
opened here https://github.com/hydrator/wrangler/pull/163

File: wrangler-service/src/main/java/co/cask/wrangler/service/s3/S3Service.java [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+  }
+
+  /**
+   * Tests S3 Connection.
+   *
+   * @param request HTTP Request handler.
+   * @param responder HTTP Response handler.
+   */
+  @POST
+  @Path("test")
```
Comment (nitinmotgi) [inline wrangler-service/src/main/java/co/cask/wrangler/service/s3/S3Service.java:105]: s3/test

File: wrangler-service/src/main/java/co/cask/wrangler/service/s3/S3Service.java [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+  }
+
+  /**
+   * Lists S3 buckets owned by the user identified by the credentials.
+   *
+   * @param request HTTP Request handler.
+   * @param responder HTTP Response handler.
+   */
+  @POST
+  @Path("buckets")
```
Comment (nitinmotgi) [inline wrangler-service/src/main/java/co/cask/wrangler/service/s3/S3Service.java:148]: s3/buckets

File: wrangler-service/src/main/java/co/cask/wrangler/service/s3/S3Service.java [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+    }
+  }
+
+  /**
+   * Lists S3 bucket's contents for the given prefix path.
+   * @param request HTTP Request handler.
+   * @param responder HTTP Response handler.
+   */
+  @POST
+  @Path("explore/buckets/{bucket-name}")
```
Comment (nitinmotgi) [inline wrangler-service/src/main/java/co/cask/wrangler/service/s3/S3Service.java:169]: s3/buckets/{bucket-name}

File: wrangler-service/src/main/java/co/cask/wrangler/service/s3/S3Service.java [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+    }
+  }
+
+  /**
+   * Loads s3 object into workspace
+   * @param request HTTP Request handler.
+   * @param responder HTTP Response handler.
+   */
+  @POST
+  @Path("load/buckets/{bucket-name}")
```
Comment (nitinmotgi) [inline wrangler-service/src/main/java/co/cask/wrangler/service/s3/S3Service.java:205]: s3/buckets/{bucket-name}

File: wrangler-service/src/main/java/co/cask/wrangler/service/s3/S3Service.java [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+        return;
+      }
+      AmazonS3 s3 = intializeAndGetS3Client(connection);
+      S3Object object = s3.getObject(new GetObjectRequest(bucketName, key));
+      if (object != null) {
+        try (InputStream inputStream = object.getObjectContent()) {
+          loadFile(responder, inputStream, object);
+        }
+      } else {
+        ServiceUtils.error(responder, "S3 Object not found");
```
Comment (nitinmotgi) [inline wrangler-service/src/main/java/co/cask/wrangler/service/s3/S3Service.java:226]: Fix error message -- think about the user.

File: wrangler-service/src/main/java/co/cask/wrangler/service/s3/S3Service.java [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+      id = ServiceUtils.generateMD5(id);
+      table.createWorkspaceMeta(id, name);
+
+      stream = new BufferedInputStream(inputStream);
+      byte[] bytes = new byte[(int)s3Object.getObjectMetadata().getContentLength() + 1];
+      stream.read(bytes);
+
+      // Set all properties and write to workspace.
+      Map<String, String> properties = new HashMap<>();
+      properties.put("bucketName", s3Object.getBucketName());
```
Comment (nitinmotgi) [inline wrangler-service/src/main/java/co/cask/wrangler/service/s3/S3Service.java:257]: bucketName = bucket.name or bucket-name

File: wrangler-service/src/main/java/co/cask/wrangler/service/s3/S3Service.java [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+      properties.put("key", s3Object.getKey());
+      properties.put(PropertyIds.CONNECTION_TYPE, ConnectionType.FILE.getType());
```
Comment (nitinmotgi) [inline wrangler-service/src/main/java/co/cask/wrangler/service/s3/S3Service.java:259]: S3.getType()

File: wrangler-service/src/main/java/co/cask/wrangler/service/s3/S3Service.java [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+      table.writeProperties(id, properties);
+
+      DataType dataType = DataType.fromString(s3Object.getObjectMetadata().getContentType());
+      dataType = dataType == null ? DataType.BINARY : dataType;
+      table.writeToWorkspace(id, WorkspaceDataset.DATA_COL, dataType, bytes);
+
+      // Preparing return response to include mandatory fields : id and name.
+      JsonArray values = new JsonArray();
+      JsonObject object = new JsonObject();
+      object.addProperty("bucketName", s3Object.getBucketName());
```
Comment (nitinmotgi) [inline wrangler-service/src/main/java/co/cask/wrangler/service/s3/S3Service.java:270]: bucket.name

**Pull request closed without merging**